File: backend/src/BMPCommerce.Application/Operations/Clientes/ClienteService.cs
using BMPCommerce.Application.Common.Abstractions;
using BMPCommerce.Application.Common.Exceptions;
using BMPCommerce.Domain.Common;
using BMPCommerce.Domain.Entities;
using BMPCommerce.Domain.Interfaces;

namespace BMPCommerce.Application.Operations.Clientes;

public class ClienteService : IClienteService
{
    private readonly IClienteRepository _clienteRepository;
    private readonly IVendaRepository _vendaRepository;
    private readonly IApplicationDbContext _dbContext;

    public ClienteService(
        IClienteRepository clienteRepository,
        IVendaRepository vendaRepository,
        IApplicationDbContext dbContext)
    {
        _clienteRepository = clienteRepository;
        _vendaRepository = vendaRepository;
        _dbContext = dbContext;
    }

    public async Task<IReadOnlyList<ClienteDto>> ListarAsync(CancellationToken cancellationToken)
    {
        var clientes = await _clienteRepository.GetAllAsync(cancellationToken);
        return clientes.Select(MapToDto).ToList();
    }

    public async Task<ClienteDto> ObterPorIdAsync(Guid id, CancellationToken cancellationToken)
    {
        var cliente = await _clienteRepository.GetByIdAsync(id, cancellationToken)
            ?? throw new NotFoundException("Cliente não encontrado.");

        return MapToDto(cliente);
    }

    public async Task<ClienteDto> CriarAsync(CriarClienteRequest request, CancellationToken cancellationToken)
    {
        var cliente = new Cliente(
            request.Nome,
            request.CpfCnpj,
            request.Telefone,
            request.Email,
            request.Cidade,
            request.Estado,
            request.Observacoes);

        _clienteRepository.Add(cliente);
        await _dbContext.SaveChangesAsync(cancellationToken);

        return MapToDto(cliente);
    }

    public async Task<ClienteDto> AtualizarAsync(Guid id, AtualizarClienteRequest request, CancellationToken cancellationToken)
    {
        var cliente = await _clienteRepository.GetByIdAsync(id, cancellationToken)
            ?? throw new NotFoundException("Cliente não encontrado.");

        cliente.Atualizar(
            request.Nome,
            request.CpfCnpj,
            request.Telefone,
            request.Email,
            request.Cidade,
            request.Estado,
            request.Observacoes,
            request.Ativo);

        await _dbContext.SaveChangesAsync(cancellationToken);

        return MapToDto(cliente);
    }

    public async Task<Result> ExcluirAsync(Guid id, CancellationToken cancellationToken)
    {
        var cliente = await _clienteRepository.GetByIdAsync(id, cancellationToken)
            ?? throw new NotFoundException("Cliente não encontrado.");

        if (await _vendaRepository.ExisteVendaComClienteAsync(id, cancellationToken))
        {
            return Result.Failure("Cliente possui vendas registradas e não pode ser excluído. Inative-o em vez de excluir.");
        }

        _clienteRepository.Remove(cliente);
        await _dbContext.SaveChangesAsync(cancellationToken);

        return Result.Success();
    }

    private static ClienteDto MapToDto(Cliente cliente) => new(
        cliente.Id,
        cliente.Nome,
        cliente.CpfCnpj,
        cliente.Telefone,
        cliente.Email,
        cliente.Cidade,
        cliente.Estado,
        cliente.Observacoes,
        cliente.Ativo,
        cliente.CreatedAt,
        cliente.UpdatedAt);
}

File: backend/app/services/cliente_service.py
"""Equivalente a Application/Operations/Clientes/ClienteService.cs."""

from __future__ import annotations

from uuid import UUID

from sqlalchemy.orm import Session

from app.core.exceptions import NotFoundException
from app.domain.cliente import Cliente
from app.domain.common import Result
from app.repositories.cliente_repository import ClienteRepository
from app.repositories.venda_repository import VendaRepository
from app.schemas.cliente import AtualizarClienteRequest, ClienteDto, CriarClienteRequest


class ClienteService:
    def __init__(self, session: Session) -> None:
        self._session = session
        self._clientes = ClienteRepository(session)
        self._vendas = VendaRepository(session)

    def listar(self) -> list[ClienteDto]:
        return [self._to_dto(c) for c in self._clientes.get_all()]

    def obter_por_id(self, cliente_id: UUID) -> ClienteDto:
        cliente = self._clientes.get_by_id(cliente_id)
        if cliente is None:
            raise NotFoundException("Cliente não encontrado.")

        return self._to_dto(cliente)

    def criar(self, request: CriarClienteRequest) -> ClienteDto:
        cliente = Cliente(
            nome=request.nome,
            cpf_cnpj=request.cpf_cnpj,
            telefone=request.telefone,
            email=request.email,
            cidade=request.cidade,
            estado=request.estado,
            observacoes=request.observacoes,
        )

        self._clientes.add(cliente)
        self._session.commit()

        return self._to_dto(cliente)

    def atualizar(self, cliente_id: UUID, request: AtualizarClienteRequest) -> ClienteDto:
        cliente = self._clientes.get_by_id(cliente_id)
        if cliente is None:
            raise NotFoundException("Cliente não encontrado.")

        cliente.atualizar(
            nome=request.nome,
            cpf_cnpj=request.cpf_cnpj,
            telefone=request.telefone,
            email=request.email,
            cidade=request.cidade,
            estado=request.estado,
            observacoes=request.observacoes,
            ativo=request.ativo,
        )

        self._clientes.update(cliente)
        self._session.commit()

        return self._to_dto(cliente)

    def excluir(self, cliente_id: UUID) -> Result[None]:
        cliente = self._clientes.get_by_id(cliente_id)
        if cliente is None:
            raise NotFoundException("Cliente não encontrado.")

        if self._vendas.existe_venda_com_cliente(cliente_id):
            return Result.failure(
                "Cliente possui vendas registradas e não pode ser excluído. Inative-o em vez de excluir."
            )

        self._clientes.remove(cliente)
        self._session.commit()

        return Result.success()

    @staticmethod
    def _to_dto(cliente: Cliente) -> ClienteDto:
        return ClienteDto.model_validate(cliente)
